### ISSUE TITLE
Custom triggers

### ISSUE THREAD

**Issue by galvesribeiro** [Azure/Azure-Functions#1154]
Hello folks!

We're building a fully serverless Event Store solution here and as part of it, we have to deal with Projections. 

Projections happens when an event is written to the Event Store so other parties can do something else (i.e. calculate a materialised view of the event stream).

So, I'm looking to create a trigger that will start a function whenever a projection has to be processed.

I've found this sample https://medium.com/@yuka1984/how-to-make-custom-trigger-with-azure-functions-3bd85551ac5e but it looks like there is no official docs on it.

Can someone point to the right directions on it?

Thanks! Really appreciate any inputs.

**Comment by galvesribeiro** [comment]
cc: @fabiocav

**Comment by soninaren** [comment]
@galvesribeiro it would be a good idea to understand the complete scenario and see if any of our existing triggers could be of use. 
cc: @jeffhollan

**Comment by jeffhollan** [comment]
In general we recommend for custom triggers you leverage Azure Event Grid which can send a Webhook.  We don't have much / any docs on potentially wrapping that event grid event in a customer trigger extension.  @mattchenderson has been looking into the documentation of extensions I believe

**Comment by galvesribeiro** [comment]
Thanks folks. In the end of the day, I guess we're ending with EventGrid + Queues to dispatch the projections.

The downside is that looks like EventGrid has serialization issues with their `Data` field...

One thing I've noticed is that EventGrid has only the trigger for AZ Func. Would be cool if we have the output binding as well.

**Comment by jeffhollan** [comment]
Incidentally was looking around the code the other day and it looks like there may be one already 

https://github.com/Azure/azure-functions-eventgrid-extension/blob/dev/src/EventGridExtension/OutputBinding/EventGridAttribute.cs

It just isn't documented.  I haven't tested it out yet but if it does work will make a note that we should document.  @brandonh-msft I seem to vaguely remember this - has this shipped in the public NuGet? Do we just need to update our docs?

**Comment by galvesribeiro** [comment]
@jeffhollan thanks! That help!

I've just added this and  it worked:

```
[EventGrid(
                TopicEndpointUri = Constants.PROJECTIONS_ENDPOINT,
                TopicKeySetting = Constants.PROJECTIONS_TOKEN)] IAsyncCollector<EventGridEvent> outputEvents
```

Indeed it is just missing the docs...

On a side question for you... Since there is a work-in-progress to make Functions more DI-friendly from @fabiocav, don't you think those bindings should be injected on the function class rather then use an attribute? Having things defined on attributes makes thing harder to unit test it...

**Comment by jeffhollan** [comment]
Will probably need to be a separate issue (feel free to open or find and link) but would be curious to learn more of how you envision it.  Not sure I fully understand what aspects could be made easier today in regards to bindings and testability - not that it's super easy today, but not sure what the easier path may look like.

**Comment by galvesribeiro** [comment]
Ok will think about it and open one...

Thanks. Should I keep this issue open for the sake of tracking the custom triggers?

**Comment by brandonh-msft** [comment]
@jeffhollan Yeah I submitted it but the publishing to nupkg needed to happen on your guys' side to have all the right signing, etc so just left that to you. Happy to sync w/ @mattchenderson on docs though - IIRC I was pretty verbose in the README so should be pretty quick.

**Comment by galvesribeiro** [comment]
@brandonh-msft I don't think you need a new package. `Microsoft.Azure.WebJobs.Extensions.EventGrid` 2.0.0 contains it. It is just missing the docs, that is why I didn't found it...

**Comment by paulbatum** [comment]
@CasperWSchmidt and @SeanFeldman  were asking in https://github.com/Azure/azure-functions-host/issues/2746 about the state of support for custom triggers, and I want to discuss that further here.

We have some ideas about how we could add support for custom triggers that would run in the [Functions Premium plan](https://docs.microsoft.com/en-us/azure/azure-functions/functions-premium-plan). The reason it would depend on the premium plan is that there is always at least one VM assigned to the plan, and it could run your activation/scaling logic. This would build on a recent feature we developed called [runtime scale monitoring](https://docs.microsoft.com/en-us/azure/azure-functions/functions-networking-options#premium-plan-with-virtual-network-triggers) that allows functions to work with event sources that are protected by networking restrictions, such as a storage account with private link enabled.

For your scenarios, would running in the premium plan be an acceptable requirement if it allowed you to implement a custom trigger?

**Comment by CasperWSchmidt** [comment]
@paulbatum according to the link in the original post and a similar guide I found, it is already possible by implementing ITriggerBinding etc. What I'm not sure about is if this works for all plans (app service/consumption/premium) and what the costs are. AFAIK the code running behind the scenes for the triggers provided by MS is free of charge - only when a function is executed you pay when using consumption plan. With a custom trigger I guess some code will have to run behind the scenes of the actual function. It makes perfect sense that this is only possible using premium or app service plan (where you can have "always on" or similar activated).

I assume that implementing ITriggerBinding etc. is also how the build-in triggers are implemented so maybe you just need some documentation (including costs and requirements) on the topic?

**Comment by paulbatum** [comment]
Indeed, ITriggerBinding is what our trigger implementations are based on and yes, you would run into problems in the consumption plan because nothing will make scaling decisions for you. 

You are correct that much of the work required here is documentation, once the scenario is scoped down to dedicated/premium. But there is some more to it - for example we would need to review the usability of the APIs we built for controlling scaling decisions from the runtime. So far it was only built for internal consumption. 

Also, it introduces a new scenario that we would need to provide commercial support for. We'd need to be confident that we can quickly assist customers when they run into problems with their custom trigger implementation running on our platform. We've learned over the years that these scenarios where there is a blend of platform code and customer code running together can be tricky. When there is a bug in a custom input binding or output binding there is an exception, a stack trace, etc. A bug in a custom trigger implementation can easily mean that your functions just dont run, and that can be a challenge to root cause.

**Comment by gorillapower** [comment]
Wanted to check back into this thread and see if there is support for custom trigger development at this time @paulbatum ? Is there support for having custom activation/scaling logic for consumption plan functions?

**Comment by paulbatum** [comment]
@gorillapower Things are relatively unchanged regarding custom triggers. There is no support for custom activation / scaling logic in consumption.

**Comment by gorillapower** [comment]
@paulbatum thank you. Im looking at the `Microsoft.Azure.WebJobs.Extensions.ServiceBus` code and see there is a class that is implementing `IScaleMonitor` and specifically this methodt that is returning the enum `ScaleVote`, presumably to inform the host to scale accordingly? https://github.com/Azure/azure-functions-servicebus-extension/blob/dev/src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusScaleMonitor.cs#L178

I also see the servicebus listener is implementing `IScaleMonitorProvider` that then gets an instance of the `IScaleMonitor`. Just to confirm, if I write a custom listener, like is in the ServiceBus Extension, that impelments these interfaces, that when deployed in Azure, the consumption host will not scale according to the custom listener and scalemonitor implementation?

**Comment by paulbatum** [comment]
Correct - in consumption it would not use your implementation. See my [comment above](https://github.com/Azure/Azure-Functions/issues/1154#issuecomment-736905558) about this.
